numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -110,7 +145,43 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
         return NPY_SUCCEED;
     }
 
-    len = PySequence_Size(obj);
+    /*
+    * If obj is a scalar we skip all the useless computations and jump to
+    * intp_from_scalar as soon as possible.
+    */
+    if (PyLong_CheckExact(obj) || !PySequence_Check(obj)) {
+        len = 1;
+
+        seq->ptr = npy_alloc_cache_dim(1);
+        if (seq->ptr == NULL) {
+            PyErr_NoMemory();
+            return NPY_FAIL;
+        }
+        seq->len = len;
+
+        npy_intp rvalue = intp_from_scalar(obj, (npy_intp *)seq->ptr, 0);
+        /*
+        * Check whether the value set by intp_from_scalar is different than
+        * the value returned, which means that an error occurred.
+        */
+        if (PyErr_Occurred() != NULL || *((npy_intp *)seq->ptr) != rvalue) {
+            npy_free_cache_dim_obj(*seq);
+            seq->ptr = NULL;
+            return NPY_FAIL;
+        }
+
+        return NPY_SUCCEED;
+    }
+
+    PyObject *seq_obj = PySequence_Fast(obj,
+        "An error occurred during a call to `PySequence_Fast`, an iterable "
+        "object is required. The given object is not a scalar integer nor a "
+        "sequence.");
```

Comment:
This error message (and also below) is not user-friendly. Maybe check what it is, and what it was.  If we end up clearing it (due to a try/except), you can leave it empty I think.  Otherwise, it should be something like `expected an integer or a sequence of integers`.

Reply:
Fixed in 12c9b4e